csharp/src/Ice/IceDiscovery/LocatorRegistry.cs

```diff
@@ -0,0 +1,283 @@
+                        return proxy;
+                    }
+                    catch
+                    {
+                        // Ignore.
+                    }
+                }
+
+                return null;
+            }
+        }
+
+        internal (IReadOnlyList<EndpointData> Endpoints, bool IsReplicaGroup) ResolveAdapterId(string adapterId)
+        {
+            lock (_mutex)
+            {
+                if (_ice2Adapters.TryGetValue(adapterId, out IReadOnlyList<EndpointData>? endpoints))
+                {
+                    return (endpoints, false);
+                }
+
+                if (_replicaGroups.TryGetValue((adapterId, Protocol.Ice2), out HashSet<string>? adapterIds))
+                {
+                    Debug.Assert(adapterIds.Count > 0);
+                    return (adapterIds.SelectMany(id => _ice2Adapters[id]).ToList(), true);
+                }
+
+                return (ImmutableArray<EndpointData>.Empty, false);
+            }
+        }
+
+        internal string ResolveWellKnownProxy(Identity identity)
+        {
+            if (identity.Name.Length == 0)
+            {
+                return "";
+            }
+
+            lock (_mutex)
+            {
+                foreach ((string replicaGroupId, Protocol protocol) in _replicaGroups.Keys)
+                {
+                    if (protocol == Protocol.Ice2)
+                    {
+                        try
+                        {
+                            // This proxy is an indirect proxy with a location (the replica group ID).
+                            IObjectPrx proxy = _dummyIce2Proxy.Clone(
+                                IObjectPrx.Factory,
+                                identity: identity,
+                                location: ImmutableArray.Create(replicaGroupId));
+                            proxy.IcePing();
+                            return replicaGroupId;
+                        }
+                        catch
+                        {
+                            // Ignore and move on to the next replica group
+                        }
+                    }
+                }
+
+                foreach (string adapterId in _ice2Adapters.Keys)
+                {
+                    try
+                    {
+                        IObjectPrx proxy = _dummyIce2Proxy.Clone(
+                            IObjectPrx.Factory,
+                            identity: identity,
+                            location: ImmutableArray.Create(adapterId));
+                        proxy.IcePing();
+                        return adapterId;
+                    }
+                    catch
+                    {
+                        // Ignore.
+                    }
+                }
+
+                return "";
+            }
+        }
+
+        private void AddAdapterToReplicaGroup(string adapterId, string replicaGroupId, Protocol protocol)
+        {
+            // Must be called with _mutex locked.
```

Comment:
Should we just lock the mutex, I thought this was preferred in C# to this comments

Reply:
The reason we use _mutex and not concurrent collections is to update both collections atomically.

So calling this method without _mutex locked is incorrect whether or not we relock _mutex in there. And locking mutex could give the impression it's fine when it's not.

Reply:
Reworked the helper methods to avoid this situation.